### PR TITLE
fix(ppo): exclude padding tokens from entropy calculation

### LINKED ISSUE #2022: Negative Entropy in TRL PPOv2Trainer TLDR Example
### System Info

- `transformers` version: 4.44.0
- Platform: Linux-5.4.0-162-generic-x86_64-with-glibc2.31
- Python version: 3.11.9
- Huggingface_hub version: 0.23.4
- Safetensors version: 0.4.3
- Accelerate version: 0.32.1
- Accelerate config:    - compute_environment: LOCAL_MACHINE
        - distributed_type: FSDP
        - mixed_precision: bf16
        - use_cpu: False
        - debug: True
        - num_processes: 2
        - machine_rank: 0
        - num_machines: 1
        - rdzv_backend: static
        - same_network: True
        - main_training_function: main
        - enable_cpu_affinity: False
        - fsdp_config: {'fsdp_activation_checkpointing': True, 'fsdp_auto_wrap_policy': 'TRANSFORMER_BASED_WRAP', 'fsdp_backward_prefetch': 'BACKWARD_PRE', 'fsdp_cpu_ram_efficient_loading': True, 'fsdp_forward_prefetch': True, 'fsdp_offload_params': True, 'fsdp_sharding_strategy': 'FULL_SHARD', 'fsdp_state_dict_type': 'SHARDED_STATE_DICT', 'fsdp_sync_module_states': True, 'fsdp_use_orig_params': True}
        - downcast_bf16: no
        - tpu_use_cluster: False
        - tpu_use_sudo: False
        - tpu_env: []
        - dynamo_config: {'dynamo_backend': 'EAGER'}
- PyTorch version (GPU?): 2.4.0+cu121 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using distributed or parallel set-up in script?: Yes
- Using GPU in script? Yes
- GPU type: NVIDIA A100-SXM4-80GB

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder
- [ ] My own task or dataset (give details below)

### Reproduction

In [TRL's PPOv2Trainer TLDR example](https://github.com/huggingface/trl/blob/main/examples/scripts/ppo/ppo_tldr.py), run the default command:

https://github.com/huggingface/trl/blob/main/examples/scripts/ppo/ppo_tldr.py#L32-L44 

### Expected behavior

Entropy must be non-negative. However, when I run the official example, the entropy is negative:

image

I don't think I'm making a mistake because this negative entropy also appears in the official documentation at https://huggingface.co/docs/trl/ppov2_trainer. Specifically, look early in training, at maybe 20k episodes:

image

The [documentation](https://huggingface.co/docs/trl/ppov2_trainer) describes `objective/entropy` as "The mean entropy of the policy, indicating the randomness of the actions chosen by the policy." If this is incorrect, and some other quantity is computed instead, then perhaps the documentation needs to be updated?

### PR DESCRIPTION
**Pull request by mukund1985**
## Problem
`objective/entropy` goes negative during PPO training, which is mathematically impossible for a discrete distribution — Shannon entropy H(p) = −Σ p·log(p) ≥ 0 always.

Root cause: `INVALID_LOGPROB = 1.0` is used as a sentinel for padding positions. The current calculation:

```python
mean_entropy = (-logprobs).sum(1).mean()
```

contributes `−1.0` per padding position, driving the sum negative.

Fixes #2022.

## Solution
Mask padding tokens before summing:

```python
mean_entropy = ((-logprobs) * (~padding_mask).float()).sum(1).mean()
```

`padding_mask` is already computed in the same scope. One-line change.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

## AI writing disclosure
- [ ] No AI usage: the PR was written entirely by a human.
- [x] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [ ] AI-generated: the PR was mostly or fully generated by an AI tool.

---

> [!NOTE]
> **Low Risk**
> Logging-only change to one metric formula; no change to gradients, rewards, or policy updates.
> 
> **Overview**
> Fixes **`objective/entropy`** logging during PPO rollouts so padded response positions no longer count toward the metric.
> 
> After padding, `logprobs` uses **`INVALID_LOGPROB` (1.0)** as a sentinel; the old **`(-logprobs).sum(1)`** treated each pad slot as **−1**, which could make logged entropy negative. The rollout metric now zeros those positions with the existing **`padding_mask`** before summing. **PPO loss and `policy/entropy_avg`** (entropy from logits in the update loop) are unchanged.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 22f1a73017b3e99d270128ebc3daf9f6c53d9ed5. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by mukund1985** [comment]
@qgallouedec @albertvillanova — would appreciate a review when you get a chance. One-line fix, low risk, happy to add tests or docs if that's blocking merge.

**Comment by qgallouedec** [comment]
PPO is currently in maintenance mode, so in theory we're no longer making changes to the trainer, but as an exception, I'm going to merge this fix because it's fairly simple and localized. Thanks!

File: trl/experimental/ppo/ppo_trainer.py [hunk 1]
```diff
@@ -880,7 +880,11 @@ def repeat_generator():
                     empty_cache()
             with torch.no_grad():
                 mean_kl = kl.sum(1).mean()
-                mean_entropy = (-logprobs).sum(1).mean()
+                # Exclude padding tokens (filled with INVALID_LOGPROB=1.0) from the entropy
+                # sum.  Without masking, each padding position contributes -1.0 to the sum,
+                # which can drive `objective/entropy` negative — impossible for a discrete
+                # distribution.  See https://github.com/huggingface/trl/issues/2022
```
Comment (qgallouedec) [inline trl/experimental/ppo/ppo_trainer.py]:
```suggestion
                # Mask padding positions, filled with INVALID_LOGPROB, else they'd drag the entropy negative
```

**Review by qgallouedec — APPROVED** [review]
